### ISSUE TITLE
some non-structurally recursive operator definitions are allowed

### ISSUE THREAD

**Issue by alleystoughton** [EasyCrypt/easycrypt#62]
According to Pierre-Yves, non-structurally recursive operator definitions like

```
op f (xs : 'a list) : int =
  with xs = []      => 0
  with xs = z :: zs => f (drop 2 zs).
```

should be rejected (at the moment, this is accepted).

**Comment by chdoc** [comment]
I'm not sure about the precise rules implemented in EC, but the definition of `f` is actually structurally recursive according to the definition used in Coq. That is, the following is accepted:
```coq
Fixpoint drop n (s : list nat) := 
  match s with
  | nil => s
  | cons _ s' => match n with
                | 0 => s
                | S n' => drop n' s'
                end
  end.

Fixpoint f (xs : list nat) :=
  match xs with 
  | nil => 0
  | cons x zs => f (drop 2 zs)
  end. 
```
The point is that all recursive calls of `drop` return either `s` or directly return the recursive call. Hence `drop 2 zs` can be recognized as structurally no-greater than `zs`. Removing this rule severely limits what one can do (conveniently) with structural recursion.

**Comment by chdoc** [comment]
@strub , is there something to be done here, or can this be closed as "working as designed"?

**Comment by strub** [comment]
At some point, we should implement a termination checker and disable the one of Why3. Let' keep this issue opened.

**Comment by chdoc** [comment]
Presumably, "at some point" is after the 2022.1 release, so I guess this can be removed from the milestone.

**Comment by strub** [comment]
Yes, you're right.

**Comment by alleystoughton** [comment]
Of course, it would be nice to document what the structural recursion rules are ;-)

**Comment by oskgo** [comment]
I celebrated too early. The following fails with `Ident Top_baz is not yet declared`:
```
op g: (nat -> nat) -> nat.

op baz (n: nat) (m: nat): nat = 
  with n = O    => O
  with n = S n' => g (baz n').

lemma test n m: baz n = baz m.
proof.
smt().
```
I have no idea how to work around this.

**Comment by strub** [comment]
IMO, this is a different issue that comes from the Why3 encoding of fixpoints.

**Comment by chdoc** [comment]
I agree, the last one is a different issue. There may be an issue with the translation and irrelevant arguments (e.g. `m` above). In particular, the definition is fine, as can be witnessed by `smt()` being able to prove the defining equations: 

```
lemma bazO m : baz O m = O.
proof. smt(). qed.

lemma bazS n m : baz (S n) m = g (baz n).
proof. smt(). qed.
```
